src/include/storage/metric/thread_level_stats_collector.h

```diff
@@ -0,0 +1,372 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <vector>
+#include "catalog/catalog_defs.h"
+#include "common/container/concurrent_map.h"
+#include "storage/metric/abstract_metric.h"
+#include "storage/metric/abstract_raw_data.h"
+#include "storage/metric/metric_defs.h"
+#include "transaction/transaction_manager.h"
+
+namespace terrier {
+
+namespace transaction {
+class TransactionContext;
+}  // namespace transaction
+
+namespace storage::metric {
+/**
+ * @brief Class responsible for collecting raw data on a single thread.
+ *
+ * Each thread should be assigned one collector, at the time of system startup, that is globally
+ * unique. This is to ensure that we can collect raw data in an non-blocking way as the
+ * collection code runs on critical query path. Periodically a dedicated aggregator thread
+ * will put the data from all collectors together into a meaningful form.
+ *
+ * @warning The constructor of this class adds itself to a map of collector. The destructor
+ * of this class erases itself from the map. The map supports concurrent insert only. This means
+ * that calling the destructor concurrently with constructors and destructors of other instances
+ * is unsafe! Therefore, ThreadLevelStatsCollector should have the same scope as the worker pool.
+ * When the worker pool resizes, the CollectorsMap needs to be resized as well.
+ */
+class ThreadLevelStatsCollector {
+ public:
+  /**
+   * Concurrent unordered map between thread ID and pointer to an instance of this class
+   */
+  using CollectorsMap = common::ConcurrentMap<std::thread::id, ThreadLevelStatsCollector *, std::hash<std::thread::id>>;
+
+  /**
+   * @return the Collector for the calling thread
+   */
+  static ThreadLevelStatsCollector *GetCollectorForThread() {
+    std::thread::id tid = std::this_thread::get_id();
+    auto const iter = collector_map_.Find(tid);
+    if (iter == collector_map_.End()) {
+      return new ThreadLevelStatsCollector();
```

Comment:
`new ThreadLevelStatsCollector()` has the side effect of adding the created `ThreadLevelStatsCollector` to the `collector_map_`. This is not clear from the usage here. Would it be better to remove it as side effect and explicitly add it to the `collector_map_` here? Or to note this as a comment?